GitUI/WaitCursorScope.cs

```diff
@@ -0,0 +1,46 @@
+﻿using System;
+using System.Windows.Forms;
+
+namespace GitUI
+{
+    /// <summary>
+    /// Sets the mouse cursor to <see cref="Cursors.WaitCursor"/> to indicate UI activity over some lexical scope.
+    /// </summary>
+    /// <remarks>
+    /// Usage is:
+    /// <code>
+    /// using (new WaitCursorScope())
+    /// {
+    ///     // perform UI activity
+    /// }
+    /// </code>
+    /// This pattern ensures:
+    /// <list type="bullet">
+    /// <item>The cursor is always restored when the block exits, whether by fall-through, return or exception.</item>
+    /// <item>Nested scopes work properly. That is, an inner scope doesn't reset the mouse when it exits.
+    /// Only the top-most scope will fully restore the mouse cursor.</item>
+    /// </list>
+    /// </remarks>
+    internal sealed class WaitCursorScope : IDisposable
```

Comment:
:bulb: `struct`

Reply:
:memo: you'll need to replace the parameterless constructor with a factory method, but it shouldn't be a problem.

Reply:
I'd considered using a struct but assumed the value would be boxed for the IDisposable call. Maybe the compiler binds directly? Any other reason this should be a struct?